### PR TITLE
Semi-Supervised SRM code

### PR DESCRIPTION
**Pull request by TuKo**
Adding the Semi-Supervised SRM code, tests and examples.

In addition, I modified the raider dataset files in the examples for SRM and SS-SRM. The SRM examples (python and notebook) were corrected accordingly.

### CODE REVIEW

**Comment by TuKo** [comment]
@cameronphchen Could you please review this PR as well? Thank you.

**Comment by cameronphchen** [comment]
@TuKo Sure, I'll do a review. Do you mind fixing the failed Travis CI first? I checked the error message, but not quite sure what's the cause.

**Comment by TuKo** [comment]
@cameronphchen, I looked at the build and didn't find any actual problem related to the code. It is failing only in one version of Mac and it is fine in the other one for Mac and Linux. The error does not seem related to the tests which is actually passing (https://travis-ci.org/IntelPNI/brainiak/jobs/161482100#L303). It is probably related to some library that we are using for the testing. I will check this with @mihaic later.

Please, go ahead and don't wait for this to be resolved. I will try to update during the day. Thanks.

**Comment by mihaic** [comment]
@cameronphchen Indeed, the Travis error is not related to this PR. Please proceed with the review.

**Comment by mihaic** [comment]
@cameronphchen Please remember about this review.

**Comment by cameronphchen** [comment]
Yes, I still remember this. Will try to finish before the Friday meeting.

File: brainiak/funcalign/sssrm.py [hunk 1]
```diff
@@ -0,0 +1,809 @@
...
+
+class SSSRM(BaseEstimator, ClassifierMixin, TransformerMixin):
+    """Semi-Supervised Shared Response Model (SS-SRM)
+
+    Given multi-subject data, factorize it as a shared response S among all
+    subjects and an orthogonal transform W per subject, using also labeled
+    data to train a Multinomial Logistic Regression (MLR) classifier (with
+    l2 regularization) in a semi-supervised manner:
+
+    .. math:: X_i \\approx W_i S ,~for~all~i=1\dots N
```
Comment (cameronphchen) [inline brainiak/funcalign/sssrm.py]: Is there a reason we write the math like this instead of the objective function? It's a bit unclear how do the parameters affect the model. In other words, there's no gamma and "cost" in the math.

File: brainiak/funcalign/sssrm.py [hunk 2]
```diff
@@ -0,0 +1,809 @@
...
+    n_iter : int, default: 10
+        Number of iterations to run the algorithm.
+
+    features : int, default: 50
+        Number of features to compute.
+
+    gamma : float, default: 1.0
+        Regularization parameter for the classifier.
+
+    cost : float, default: 0.5
```
Comment (cameronphchen) [inline brainiak/funcalign/sssrm.py]: can we not use "cost" as the name? This name is a bit confusing.... especially this is just a parameter not really a cost. We use alpha in the paper, can we also use alpha here?

File: brainiak/funcalign/sssrm.py [hunk 3]
```diff
@@ -0,0 +1,809 @@
...
+        for m in range(len(w)):
+            s = s + w[m].T.dot(data[m])
+        s /= len(w)
+        return s
+
+    def _objective_function(self, data_align, data_sup, labels, w, s, theta,
+                            bias):
+        """Compute the objective function of the Semi-Supervised SRM
+
+        .. math:: (1-C)*Loss_{SRM}(W_i,S;X_i)
```
Comment (cameronphchen) [inline brainiak/funcalign/sssrm.py]: same here, can we use notations that are consistent to the paper?

File: brainiak/utils/utils.py [hunk 4]
```diff
@@ -105,6 +105,71 @@ def fast_inv(a):
         raise
 
 
+def sumexp_stable(data):
```
Comment (cameronphchen) [inline brainiak/utils/utils.py]: why do we need this instead of a softmax function directly? are we going to use this besides the calculation in softmax?

File: brainiak/funcalign/sssrm.py (hunk 2 above)
Comment (TuKo) [inline brainiak/funcalign/sssrm.py]: done

File: brainiak/funcalign/sssrm.py (hunk 3 above)
Comment (TuKo) [inline brainiak/funcalign/sssrm.py]: done

File: brainiak/funcalign/sssrm.py (hunk 1 above)
Comment (TuKo) [inline brainiak/funcalign/sssrm.py]: done

File: brainiak/utils/utils.py (hunk 4 above)
Comment (TuKo) [inline brainiak/utils/utils.py]: We are using this for the objective function computation. This is only for logging.

**Comment by mihaic** [comment]
@cameronphchen To test on the whole Raiders data we need long-running tests, which are different from the short-running tests we run now for PRs. We have an issue for documenting acceptable runtimes for the existing tests (issue #107) and an issue for long-running tests (issue #123). We can discuss those issues, but I think we should not hold this PR until they are solved; if you would like, you can create an SSSRM-specific testing issue blocked by issue #123.

**Comment by cameronphchen** [comment]
@mihaic Definitely, that shouldn't be blocking this PR.

**Comment by TuKo** [comment]
Jenkins, retest this please.

**Comment by TuKo** [comment]
@cameronphchen , please review the latest changes. If everything is fine, please approve the PR. Otherwise let me know and I will add changes.
(Linux testing is failing because there is some issue with the system)

**Comment by mihaic** [comment]
Jenkins, retest this please.

File: tests/utils/test_utils.py [hunk 5]
```diff
@@ -44,6 +44,32 @@ def test_fast_inv():
...
+    from brainiak.utils.utils import concatenate_list
+    import numpy as np
+    l = [None] * 5
+
+    l[1] = np.array([0, 1, 2])
+    l[3] = np.array([3, 4])
+    try:
+        r = concatenate_list(l, axis=0)
+    except:
+        assert True, "Could not concatenate a list"
```
Comment (mihaic) [inline tests/utils/test_utils.py]: Use `pytest.raises`, please.
